### ISSUE TITLE
Event Sheet style visual scripting

### ISSUE THREAD

**Issue by blurymind** [godotengine/godot-visual-script#27]
NOTE:
Here is the test repo that I added the prototype project to
https://github.com/blurymind/Godot-eventSheetPrototype
Everyone is free to fork or do pull requests 👍 

IDEA:
We currently have a visual scripting system similar to blueprints in Unreal - connecting nodes.
The proposal here is for a second visual scripting system, that is similar to event sheets in Construct 2 (proprietary), Multimedia fusion (proprietary) and Gdevelop (open source)
![11](https://user-images.githubusercontent.com/6495061/37964468-81886d54-31b9-11e8-9ba5-555123ad1fc7.jpg)
![GD-clickteamesque-additionOfEvents](https://user-images.githubusercontent.com/6495061/78575488-57c5f100-7823-11ea-82f8-c795ac10f98f.gif)

It is a very different approach from the one with blueprints and people learning to program are still requesting it on facebook and other godot community forums

What is an event sheet visual script in the other engines:
https://www.scirra.com/manual/44/event-sheet-view
The event sheet is pretty much a spreadsheet with two columns - a conditions column and an actions column. Both columns can be populated with logic blocks from nodes and their children that the sheet is attached to (node methods). On the left column the user can only attach conditional methods, on the right - only action methods. This clear divide makes for a very easy to learn way of setting game logic. 
On top of that the user can use expressions in both columns - so potentially use gdscript for more specific instructions. 

Rows can be nested under other rows (called sub-events), can be commented, disabled or re-enabled (just like commenting code) 
https://www.scirra.com/manual/128/sub-events
![subeventexample](https://user-images.githubusercontent.com/6495061/37964565-da595506-31b9-11e8-8d9c-03f7f1944307.png)
Some actions/condition blocks can be negated

Functions that can take parameters can be used as well, by using a special function condition block and nesting conditions/actions under its row
![image28](https://user-images.githubusercontent.com/6495061/37964431-5b118958-31b9-11e8-9a81-2a56ac99c88a.png)
![modifiedcheckmatches](https://user-images.githubusercontent.com/6495061/37964409-4a2fafa2-31b9-11e8-8fee-ac374bb4996a.png)

So What are the advantages over our current visual script:
- It is an alternative style of visual scripting with a big following - a quick look at the number of users of construct 2 and the number of succesful 2d games made with clickteam fusion should be a good enough indicator. It has also been requested on Unity forums too
https://feedback.unity3d.com/suggestions/create-a-visual-scripting-for-begginer-like-construct-2-or-behavior-machine
https://feedback.unity3d.com/suggestions/visual-scripting-based-in-event-sheets-as-construct-2

- It is easier to learn and arguably clearer for non-programmers
- An event sheet can pack much more information of the game logic on one screen than blueprints - less scrolling and panning to get the information. Less wasted empty space between logic blocks. You can technically just take a screenshot of an event sheet and share it to show someone else how to do something.
![6708 image_2e2b4e43](https://user-images.githubusercontent.com/6495061/37964332-fc3ed1a6-31b8-11e8-82d8-835ea0c2ce00.png)

- It is easier to transition learning from event sheets to scripting - because it's more similar to scripting    than blueprints

- Why is it easier to learn than blueprints - the clear divide of condition and action and obvious order of execution. The user gets a filtered list of things to do on the two columns.

- Logic blocks are easy to quickly read/find because they have icons. Most nodes in godot also have icons - they can be reused for the event sheet implementation 

- Less clicks needed to get things working - no need to connect nodes or move nodes on the blueprint table.  You just add or drop logic blocks in the cells. No need to pan at all- you only scroll and its much less.

In any case, I am writing this proposal not to say that one system is better than the other - but more with the hope of sparking interest in developing an alternative to our custom visual scripting approach - an alternative that is popular amongst people learning to code and that is a great transition to gdscript - as I found out from first hand experience 

**Addon progress report 0**

Here is a crude mockup so far:
![capture](https://user-images.githubusercontent.com/6495061/38088809-13d17092-3355-11e8-89b3-563b2015686b.PNG)

Demos of event sheet style systems that you can try online(no log in required):
https://editor.gdevelop-app.com/
https://editor.construct.net/

Event Sheet System Possible Structure:
```
|_Event sheet established variables and connections tab
|_Event sheet script tab
  |_Function(built in or custom)
      |_event sheet row (can be parented to another event sheet row or an event sheet group)
          |_Actions column
               |_Action cell (richtex label) (click to open another window to edit)
          |_ Conditions Column
               |_Condition Cell (richtex label)(click to open another window to edit)
|_Action/Condition Cell Expression Editor
  |_Gdscript editor instance - to be used for expressions
  |_Easy Click interface to access the available subnodes - their nodepaths and methods- clicks bring up menu that populates the expression editor - similar to Clickteam Fusion's
```

Inner workflow:
Event sheet resource can be attached to node -->on runtime it generates a gdscript and that is used by the game

**Addon progress report 1**

I did some work on the addon's most important building block- the event sheet cell

![es-adding](https://user-images.githubusercontent.com/6495061/39065812-e6bc306a-44ca-11e8-9053-c64369d95aee.gif)

Some background in what it does - Basically the event sheet is made out of cells. A cell can contain either conditions (getters/expressions) or actions (setters that can be fed with getters/expressions).
On the GUI side, the event cell is created via a richtextlabel node and bbcode that is generated from gdscript. When you double click on it, the richtextlabel turns into an editbox node containing the actual gdscript expression. 

So an event sheet cell has 2 modes:
- edit mode - textEdit node that can be typed into with gdscript. 
The only difference is that the user does not need to type in If/else/while - that is context sensitive to the parent container as seen in the screenshot. Every line is a new condition, so if the user hasnt started the line with "or" or somethind else, the parser automatically knows that a new line has "and" pretext

When clicked away, switches to view mode.
- view mode - richtext label - When switching to view mode, a bbCode gets parsed from the gdscript that is in edit mode and presented via an interactive richtext label. Apart of being interactive and easy to change , the goal is to present the gdscript code in a clearer way. This means showing the node's name and icon only (not the entire path) and getting rid of some words, when its obvious (such as get_ and set_). Since every clickable part is actually a url tag, when hovering over a node name for example - the user can get some information, such as the full path of the node.
 
About the Add new condition/Action menu:
This is what creates a new gdscript line of code for a condition or an action. Whats great about it is that it lets you easily browse all of the nodes inside a scene and their methods - it sort of works in an inverted way to how autocompletion works in gdscript's editor. It shows all nodes and all of their setter, getter or both methods. You can then narrow down to what you want via a filter. 
If callend from a condition cell, this menu shows only getters, if called from an actions cell, it shows both setter and getter methods.

Note that this is still full of bugs and not even half complete to share, but hopefully makes it clearer what I am proposing

**Progress report 2**
Made some planning on how it can work. Also looked into what needs to be refactored before presenting the concept addon

I made this flowchart to explain how it works at the moment
https://www.draw.io/#Hblurymind%2FGodot-eventSheetPrototype%2Fmaster%2FEventSheetDiagramPlan.xml
![eventsheetmockupplan](https://user-images.githubusercontent.com/6495061/40834668-5c1f4e58-6589-11e8-8e53-ee90d8c36e78.PNG)
Decided to refactor it to generate typed gdscript
 godotengine/godot#19264
Its much easier to create bbcode links for further helper menus when its typed

**Comment by Calinou** [comment]
> Is this worth reopening now that Godot has decided to remove Visual Scripting in 4.0?
Seems like part of the argument against event sheets was because visual scripting already existed- so it'd be good to get an update on people's opinions now that visual scripting is going away.

Note that visual script development won't be happening in core (especially for 4.0) – future development is being done in an extension, hence this issue being transferred.

**Comment by DrZanuff** [comment]
@Calinou Yeah, I know. This is much better, because we'll have more flexibility and it won't drag core development. We'll be able to create and test different ideas.

**Comment by repkino** [comment]
As an artist, I support. Event sheets give you a quick start.

**Comment by AnidemDex** [comment]
I'm still working on a different approach for visual scripting in godot (event sheet GDevelop like).

Anyway, I consider that this should be moved to a discussion (in discussions board) or to a proposal (in the proposal repository), and link this issue there.

**Comment by AnidemDex** [comment]
This is my log so far.
### tl;dr
I wanted a kind of scripting language for event sheet integration (just as it should be I think?), but since is hard to  do in 3.x (without doing magic tricks that may imply recompilation) I'll take using a resource wrapper as an approach.

This approach is similar to what I've doing in EventSystem. Sequential functions are called according to certain conditions, so you can add custom functions without messing with the core implementation. 

This approach creates a gdscript script with a certain _(very specific)_ template that can be applied to _any_ `Object` (as you do normally with scripts), not only nodes.

What I try to do is to create a very solid structure to use it as a proper script language in 4.0 _(because I didn't figure out how to use PluginScript in 3.x)_, but that can be usable in 3.x, organized to not confuse users and easily extensible to create custom event behaviours and functions.

-----
I took what I did in EventSystem as base (you can think on it as a just a single event that calls many actions with no conditions except "`if action`" that acts as "sub-event") and applying what event manager were doing for all types of objects instead. I used structures suggested by OP too.

I wanted that anyone can create an script and attach it in any node so I needed to create a kind of integration to tell godot to recognize my "custom language", but after expending many days on it and just making it be able to recognize the extension but not the implementation frustrated me; so I decided to take a different approach, avoiding to re-write the script itself (no code generation as example repo on the issue).

I created a template with gdscript that can be applied to any object, any attempt to edit an script that has this template applied will trigger a plugin, and the plugin will show a custom editor for that "script" **in the script editor tab**; easy, right? But is not as easy as it sounds.

The script is converted to a custom resource for editting, and the script reference is saved. It needs a certain context to tell the editor what is supposed to be looking at, and this context is saved in the script(now being a different resource) aswell. For nodes, this context is gather from the current scene where script owner is, for other object types I rely on other magic tricks (like saving the last edited object and verifying if that object holds that script).

In-game, the script template will use the generated resource to do the actions. Many notifications are used for this purpose, and signals were connected using the resource context, signals connected through editor will need an extra step to trigger the event. 

Most of the benefits of the eventsheet in Construct or GDevelop are only useful and applicable in objects that has a loop in godot (nodes, using process function), so I decided to make my own loop mode for all objects (very useful if you want the events happens faster than process, tied to a timer or in a manual configuration), but this is for an advanced scenario, is not relevant rn.

All of the mentioned things are implemented manually, so technically its "core" is working.

Now I'm working on the editor part. This is a little mockup (I'm a terrible designer, sorry), I'm replicating a combination of Construct and GDevelop event sheet editor.
![mock](https://user-images.githubusercontent.com/7025991/188517038-48e8e603-7966-41da-9e34-2d30aa246099.png)
1. Is the main editor view, events are added and displayed there.
2. I want to be able to split functions in different tabs to keep things organized, but also being able to keep them all in the main tab, so I just noted it here.
3. Is the base class which the script extends. Since the base script extends from object, this data is extracted from owner somehow, or hidden if there's no data about this.

I'll share code when I have something working that doesn't require magic tricks in the user side to make it work. In the meantime maybe i can comment here the progress?

**Comment by AnidemDex** [comment]
No, it'll have an implementation in 3.x that will serve as a base structure
for the language, and I aim to move that implementation to an
ScriptLanguage in 4.0

**Comment by prominentdetail** [comment]
@AnidemDex Can you describe the workflow of your eventsheet?
If it extends from an object node, does this mean that the script/eventsheet is limited to the object node? I think that would defeat the purpose of using eventsheets. Having an event sheet limited to one seems restrictive. It'd be better if you can use any type of node/object in the eventsheet.

**Comment by AnidemDex** [comment]
Object is the base type for anything that you can extend.
In 3.x it needs to be this way to be able to apply the script template in
anything.

The script template is the hearth of the event sheet, the information about
what the event sheet does is saved in a resource. The script uses this
resource to apply the event sheet logic

**Comment by prominentdetail** [comment]
@AnidemDex How does the workflow work though? From the perspective of the end-user, how do you use it?

**Comment by AnidemDex** [comment]
@prominentdetail It should be similar enough to what you normally do with other scripts. You open an script that contains the template, and the plugin will take care of showing the event sheet editor. The only extra step is to select the script template in new scripts that you want to apply event sheet style to.

Things that can done when you're in the event sheet editor are planned to be similar enough to the tools that gdevelop and construct has (like the add event/condition/action buttons, a menu to filter actions and conditions by categories on selection, comments, functions and groups)

**Comment by eon-s** [comment]
I wonder if could be better to open a discussion (that you can update) in proposals' repository so it reaches more people, at least until the project adds a better repository/site/something to talk about and request addons.

**Comment by St3fanos** [comment]
> This is my log so far.
> ### tl;dr
> 
> I wanted a kind of scripting language for event sheet integration (just as it should be I think?), but since is hard to do in 3.x (without doing magic tricks that may imply recompilation) I'll take using a resource wrapper as an approach.
> 
> This approach is similar to what I've doing in EventSystem. Sequential functions are called according to certain conditions, so you can add custom functions without messing with the core implementation.
> 
> This approach creates a gdscript script with a certain _(very specific)_ template that can be applied to _any_ `Object` (as you do normally with scripts), not only nodes.
> 
> What I try to do is to create a very solid structure to use it as a proper script language in 4.0 _(because I didn't figure out how to use PluginScript in 3.x)_, but that can be usable in 3.x, organized to not confuse users and easily extensible to create custom event behaviours and functions.
> 
> I took what I did in EventSystem as base (you can think on it as a just a single event that calls many actions with no conditions except "`if action`" that acts as "sub-event") and applying what event manager were doing for all types of objects instead. I used structures suggested by OP too.
> 
> I wanted that anyone can create an script and attach it in any node so I needed to create a kind of integration to tell godot to recognize my "custom language", but after expending many days on it and just making it be able to recognize the extension but not the implementation frustrated me; so I decided to take a different approach, avoiding to re-write the script itself (no code generation as example repo on the issue).
> 
> I created a template with gdscript that can be applied to any object, any attempt to edit an script that has this template applied will trigger a plugin, and the plugin will show a custom editor for that "script" **in the script editor tab**; easy, right? But is not as easy as it sounds.
> 
> The script is converted to a custom resource for editting, and the script reference is saved. It needs a certain context to tell the editor what is supposed to be looking at, and this context is saved in the script(now being a different resource) aswell. For nodes, this context is gather from the current scene where script owner is, for other object types I rely on other magic tricks (like saving the last edited object and verifying if that object holds that script).
> 
> In-game, the script template will use the generated resource to do the actions. Many notifications are used for this purpose, and signals were connected using the resource context, signals connected through editor will need an extra step to trigger the event.
> 
> Most of the benefits of the eventsheet in Construct or GDevelop are only useful and applicable in objects that has a loop in godot (nodes, using process function), so I decided to make my own loop mode for all objects (very useful if you want the events happens faster than process, tied to a timer or in a manual configuration), but this is for an advanced scenario, is not relevant rn.
> 
> All of the mentioned things are implemented manually, so technically its "core" is working.
> 
> Now I'm working on the editor part. This is a little mockup (I'm a terrible designer, sorry), I'm replicating a combination of Construct and GDevelop event sheet editor. ![mock](https://user-images.githubusercontent.com/7025991/188517038-48e8e603-7966-41da-9e34-2d30aa246099.png)
> 
>     1. Is the main editor view, events are added and displayed there.
> 
>     2. I want to be able to split functions in different tabs to keep things organized, but also being able to keep them all in the main tab, so I just noted it here.
> 
>     3. Is the base class which the script extends. Since the base script extends from object, this data is extracted from owner somehow, or hidden if there's no data about this.
> 
> 
> I'll share code when I have something working that doesn't require magic tricks in the user side to make it work. In the meantime maybe i can comment here the progress?

I would happily jump to Godot If something like this existed. Waiting years for a proper engine to make a good VS system like Construct's or Gd. Everyone seems to think that blueprints like VS is the best way to do it and I don't understand the reason. 

To be honest I know why. It only got popular because 3d artists are already used to this spaghetti node bs system for shaders.
But actually helping you make a proper complete game, Being visually clear, and able to help transition to actual code only event type VS is able to do.

**Comment by blurymind** [comment]
this is now very old, but I think Gamesfromscratch recently made a very good point about what the main advantage of the event sheet visual programming is - even for professional programmers:

https://www.youtube.com/watch?v=OUNmRLvB6KI
It just works really well with a touch interface!

@groud @Calinou @swarnimarun 
We can now run godot on mobile devices, but without a keyboard it is practically useless.

Maybe that is the thing that will motivate godot developers to take another try at a visual programming approach - hopefully learning a ton from the previous scrapped node based approach, which ultimately not enough people used to maintain.

To me something that enables coding on the go would immensely increase the value of godot as a game engine on a mobile device. In terms of UX Gdevelop is absolutely the better choice atm. Having to plug a keyboard to use it creates more friction to pick it up on an android tablet.

Dont get me wrong, it absolutely is a win on Chome laptops! Even tablets with a keyboard, but imagine if you didnt even need to plug a keyboard, on top of giving a nice introduction/wrapper to gdscript to people who are still learning programming.

To me that is kinda what made visual programming in gamemaker cool - it transpiles to gml

**Comment by AnidemDex** [comment]
People are still trying to make their own visual scripting approach for Godot, see for example:

- [this post on X](https://x.com/henrique_dart/status/1713554033623834686?s=20):
![image](https://github.com/godotengine/godot-visual-script/assets/7025991/6cca451f-bc76-4cf6-a666-167a45f8b86b)

It generates GDScript code under the hood.

- [Orchestator](https://github.com/Vahera/godot-orchestrator)
![image](https://github.com/godotengine/godot-visual-script/assets/7025991/6e3ae17a-5b87-44f6-8c5f-79730a48d0de)

Haven't seen its workflow due its documentation page being down

- [Blockflow](https://github.com/AnidemDex/Blockflow)
![image](https://github.com/godotengine/godot-visual-script/assets/7025991/16479046-7b1b-4558-bf36-c764e1b173c4)

Groups of predefined (and extendable) instructions executes actions in the scene.
_(doing a little spam)_

---

I think the only reason that it was removed from core with the first approach (Godot Visual Scripting) was because how hard was to maintain compared to the benefit that it gave to users with its direct alternative GDScript. 

I know it would be hard to integrate to core (specially now, I don't think anyone is going to write a bunch of C++ modules for this, and main maintainers had a lot of work to do for engine issues), but if you are still interested about making this, you should consider making it a plugin first _(and if you do, reach me on discord if you need help, I love these kind of projects)_

**Comment by blurymind** [comment]
all of these are node based which is unfortunate :(

To be fair making an event sheet one - with ability to do expressions in gdscript is probably very difficult. I tried and kept running into limitations

**Comment by AnidemDex** [comment]
You actually don't need to generate GDScript as the initial concept was, just define the "interpreter" with gdscript (or your favorite language) :thinking:

**Comment by pietru2004** [comment]
I got referenced here from DC, Cause I have yet another idea how to add VS Scripting
![obraz](https://github.com/godotengine/godot-visual-script/assets/51274570/c194ef35-652d-49ec-a59b-b46682d1d9a6)
Basically have functions displayed as graph blocks and have ability to edit functions in side view.
Under the hood it just parses gdscript to find functions definitions and place them as blocks. Also this one is meant to help in cases of having too many functions in scripts to see them better by just helping you focus on one, also if you save while there are 2 functions defined it just adds new graph node to graph edit.
![obraz](https://github.com/godotengine/godot-visual-script/assets/51274570/29bf8a70-232e-4309-92fa-e1a75f146091)
Only 2 things missing in here are ability to reload code/notify of code change and ability to have gdscript code interpreter work in code edit... IDK how to make it work...


I hope I didn't misunderstand issue/post...

**Comment by starry-abyss** [comment]
I like the idea of tables or trees + transpiling them to GDScript under the hood. That's how scripting in Warcraft 3 was done (with their own scripting language instead of GDScript of course).

I think what a lot programmers miss is that such visual scripting languages completely remove the layer of compiler yelling at the user. It's just impossible to make mistakes that are later caught by the compiler, because the UI checks all the names and types and just won't allow user to make those mistakes in the first place. As in, you can't input a value of wrong type at all.

**Comment by Raiguri** [comment]
Someone made plugin for Godot 4 that adds Construct-like event sheets
https://t.me/wladekhack/60

**Comment by DrZanuff** [comment]
@Raiguri any chance they have a link other than Telegram?

**Comment by Raiguri** [comment]
> @Raiguri any chance they have a link other than Telegram?

~I don't think so, he didn't upload it anywhere, except his telegram.~
Oi, my bad he uploaded it to [mega](https://mega.nz/file/VCtU0CLR#AbN2kn_wgPtRwL_7hFDg6WNUNLLmA0oBYWtJG9KYQA4)

Translation of his telegram post:

> ![зображення](https://github.com/user-attachments/assets/e692103d-9693-4e9d-8777-7b815b4846d6)
> Implemented a plugin for Godot 4 that adds a event sheets as in Construct 3.
> 
> The plugin is just a visual shell with a little functionality, but it can't execute the code itself when the game starts. It has a basic implementation of code execution from the event table, but it is very simple. If there are experts in Godot 4 and willing to refine this plugin I am ready to share the project (source code).
> 
> In my opinion - the full implementation of such a plugin can be a good competition to Construct 3 and open more opportunities for development than in Construct 3.
> 
> There are 3 plugins in the addons folder:
> - event_sheet - this is the main plugin.
> - plugin_refresher - plugin for development, it is needed to update the main plugin without restarting the project.
> - explore-editor-theme - plugin for development, it helps to find out the names of standard Godot 4 icons and theme colours.
> 
> The folder ‘demo_project’ stores demo scene and object.
> 
> The folder ‘event_sheet’ stores the saves of the event table.
> 
> 
> Source code: download (https://mega.nz/file/VCtU0CLR#AbN2kn_wgPtRwL_7hFDg6WNUNLLmA0oBYWtJG9KYQA4)

**Comment by DrZanuff** [comment]
@Raiguri does it have a repo? Or they intent to make this something paid/closed source. As I don't have access to Telegram I kind of blind here. If you could share any news here it would be nice.

**Comment by Raiguri** [comment]
> @Raiguri does it have a repo? Or they intent to make this something paid/closed source. As I don't have access to Telegram I kind of blind here. If you could share any news here it would be nice.

He said in his post:
> If there are experts in Godot 4 and willing to refine this plugin I am ready to share the project (source code).

**Comment by DrZanuff** [comment]
Hope this gains some traction. Thanks for sharing.

**Comment by WladekProd** [comment]
> > @Raiguri any chance they have a link other than Telegram?
> 
> ~I don't think so, he didn't upload it anywhere, except his telegram.~ Oi, my bad he uploaded it to [mega](https://mega.nz/file/VCtU0CLR#AbN2kn_wgPtRwL_7hFDg6WNUNLLmA0oBYWtJG9KYQA4)
> 
> Translation of his telegram post:
> 
> > ![зображення](https://private-user-images.githubusercontent.com/50903186/377198868-e692103d-9693-4e9d-8777-7b815b4846d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxMDk3OTgsIm5iZiI6MTcyOTEwOTQ5OCwicGF0aCI6Ii81MDkwMzE4Ni8zNzcxOTg4NjgtZTY5MjEwM2QtOTY5My00ZTlkLTg3NzctN2I4MTViNDg0NmQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE2VDIwMTEzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxZjlhNGM4YjM3MzdjMzcxYzdlOWVmNmQyMGY2YmQzOGIzM2I2ODRlYjZjZWZiNjZjOGY1YzJlNzhjMjBmOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-d377KNUGYMbw6-6dE6xbO5KbY8gKZIMrxYVnydpD40)
> > Implemented a plugin for Godot 4 that adds a event sheets as in Construct 3.
> > The plugin is just a visual shell with a little functionality, but it can't execute the code itself when the game starts. It has a basic implementation of code execution from the event table, but it is very simple. If there are experts in Godot 4 and willing to refine this plugin I am ready to share the project (source code).
> > In my opinion - the full implementation of such a plugin can be a good competition to Construct 3 and open more opportunities for development than in Construct 3.
> > There are 3 plugins in the addons folder:
> > 
> > * event_sheet - this is the main plugin.
> > * plugin_refresher - plugin for development, it is needed to update the main plugin without restarting the project.
> > * explore-editor-theme - plugin for development, it helps to find out the names of standard Godot 4 icons and theme colours.
> > 
> > The folder ‘demo_project’ stores demo scene and object.
> > The folder ‘event_sheet’ stores the saves of the event table.
> > Source code: download (https://mega.nz/file/VCtU0CLR#AbN2kn_wgPtRwL_7hFDg6WNUNLLmA0oBYWtJG9KYQA4)

https://github.com/WladekProd/EventSheet

**Comment by AnidemDex** [comment]
@WladekProd the repository is private

**Comment by WladekProd** [comment]
> @WladekProd the repository is private

Fixed

**Comment by St3fanos** [comment]
> > > @Raiguri any chance they have a link other than Telegram?
> > 
> > 
> > ~I don't think so, he didn't upload it anywhere, except his telegram.~ Oi, my bad he uploaded it to [mega](https://mega.nz/file/VCtU0CLR#AbN2kn_wgPtRwL_7hFDg6WNUNLLmA0oBYWtJG9KYQA4)
> > Translation of his telegram post:
> > > ![зображення](https://private-user-images.githubusercontent.com/50903186/377198868-e692103d-9693-4e9d-8777-7b815b4846d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxMDk3OTgsIm5iZiI6MTcyOTEwOTQ5OCwicGF0aCI6Ii81MDkwMzE4Ni8zNzcxOTg4NjgtZTY5MjEwM2QtOTY5My00ZTlkLTg3NzctN2I4MTViNDg0NmQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE2VDIwMTEzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxZjlhNGM4YjM3MzdjMzcxYzdlOWVmNmQyMGY2YmQzOGIzM2I2ODRlYjZjZWZiNjZjOGY1YzJlNzhjMjBmOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-d377KNUGYMbw6-6dE6xbO5KbY8gKZIMrxYVnydpD40)
> > > Implemented a plugin for Godot 4 that adds a event sheets as in Construct 3.
> > > The plugin is just a visual shell with a little functionality, but it can't execute the code itself when the game starts. It has a basic implementation of code execution from the event table, but it is very simple. If there are experts in Godot 4 and willing to refine this plugin I am ready to share the project (source code).
> > > In my opinion - the full implementation of such a plugin can be a good competition to Construct 3 and open more opportunities for development than in Construct 3.
> > > There are 3 plugins in the addons folder:
> > > 
> > > * event_sheet - this is the main plugin.
> > > * plugin_refresher - plugin for development, it is needed to update the main plugin without restarting the project.
> > > * explore-editor-theme - plugin for development, it helps to find out the names of standard Godot 4 icons and theme colours.
> > > 
> > > The folder ‘demo_project’ stores demo scene and object.
> > > The folder ‘event_sheet’ stores the saves of the event table.
> > > Source code: download (https://mega.nz/file/VCtU0CLR#AbN2kn_wgPtRwL_7hFDg6WNUNLLmA0oBYWtJG9KYQA4)
> 
> https://github.com/WladekProd/EventSheet

hopefully this takes off, seems great!! fingers crossed

**Comment by blurymind** [comment]
one thing I found challenging when trying to do a proof of concept ages ago was getting expressions fields with gdscript autocompletion to work.

They are quite an important element of both construct and gdevelop when it comes to the event sheet.

How does one even do a mini input field similar to a textarea alement - with syntax autocompletion, coloring that has both gdscript and custom context mixed. Expressions need to indicate when wrong (linting) - just the same as the code editor works. You need to basically create many usable mini gdscript editors - customised to know if they are expressions for condition or action or ones supposed to return a type of value (boolen, number, string, etc) and tell you when they are not returning that type of value. They need to have correct validation in place depending on the context they are used.

Then you need to convert all of that to actual gdscript when building/playtesting - thats another huge hurdle

**Comment by AnidemDex** [comment]
I still don't like the whole idea of converting that behavior into gdscript, that would limit the capability of the system itself just trying to preserve the 1:1 scripting to gdscript.

That said, Godot now has a code edit node, that includes autocompletion, I guess it would be easy to implement gdscript completion if somehow you got access to the script server  🤔

**Comment by blurymind** [comment]
In my opinion this should try to be as close as possible to the  engine's functionality  and not attempt to add any special abstractions to the runtime.

That way:
- less likely to loose compatibility when godot updates - easier to maintain and less likely to become a dead project
- fewer gotchas when something in your game doesnt work - is it my event sheet or my gdscript
- better learning transition for anyone starting with the event sheet to moving to gdscript - once you use expresions, you can apply that knowledge to gdscript
- very much like how game maker does it - their visual programming system transitions to gml on the fly - it uses gml under the hood
- godot is not architected like gdevelop and construct from the ground up to be event sheet first engine. It is closer to game maker in that regard - game maker has its own scripting language (gml) and a UI coding frontend to it

The event sheet system can aid the developer the same way the other engines do it:
- visual hints (such as object icons) in the code
- scaffolding that gives simple representation of the code and very good overview of the available methods and available objects that can be affected
- gui aid when building an expression - giving you good overview of whats available for use in said expression - Construct, Gdevelop both took that idea from Clickteam fusion btw

Why object icons are good - they are almost like emoji representations of your nodes and game objects - when you look at them its much faster to read at a glance than their word representation. For generic types - you can completely remove the word representation and eliminate alot of noise in how logic is represented on the screen

**Comment by WladekProd** [comment]
At the moment I am rewriting the plugin completely, I decided to change the code completely to achieve easier extensibility and get rid of crutches of the old code. If there will be success - will report.

I rewrote the compiler of the event table according to the new scheme: Event Sheet -> GDScript -> Run Code.

That is, first you build the logic in the Event Sheet and when you start the project the engine creates a GD script from it and runs it. Also decided to implement adding Event Sheet for any scene instead of one generic one. So far, there's been a little success. After completing the logical part, I will start the visual part.

**Comment by WladekProd** [comment]
> If there will be success - will report.

![image](https://github.com/user-attachments/assets/f823a2c4-cc21-4839-8407-1d9509e2c573)

Generating code from the event table works even better than I imagined. There is no visual part of the event table yet, but I will start working on that once the tests are complete.

**Comment by DrZanuff** [comment]
Nice!

**Comment by AnidemDex** [comment]
> Godot now has a code edit node, that includes autocompletion, I guess it would be easy to implement gdscript completion if somehow you got access to the script server (https://github.com/godotengine/godot-visual-script/issues/27#issuecomment-2419452182)

Mandatory mention to: https://github.com/godotengine/godot/pull/95764

**Comment by WladekProd** [comment]
![image](https://github.com/user-attachments/assets/936d2fcf-ec66-488b-acef-12fc130c681c)

https://github.com/user-attachments/assets/c7fb378c-c2eb-45d4-8052-8ec6de7805f1


**Innovations:**  
- Minor bug fixes to the Event and Event Selection window (now when pressing “back” and “next” buttons, previously entered parameters will be saved for each Event and Event).
- Adding events and events to Event Sheet (all this works recursively, my brain was boiling while I was dealing with it - but I'm happy with the result).
- Sub Events system (yes, you heard me, now you can create sub-blocks, move them, change order and so on).

**Comment by DrZanuff** [comment]
Very nice man! I know maybe it is too early, but how do plan to handle created objects and pick then aftewards in conditions?

**Comment by WladekProd** [comment]
> Very nice man! I know maybe it is too early, but how do plan to handle created objects and pick then aftewards in conditions?

I'm still thinking about it, but the first draft is already there and it looks like this.

https://github.com/user-attachments/assets/a5a68c3c-9d4c-45d3-8567-9fba7e158b60

**Comment by DrZanuff** [comment]
I see what you’re aiming for here, but I think it could become too cluttered when dealing with a large number of nodes and nested structures.

Perhaps each node in the tree could have a checkbox in the inspector to mark it as "pickable" by the eventsheet? Additionally, objects created by the eventsheet could have this option enabled by default. Just brainstorming ways to handle potential challenges.

**Comment by WladekProd** [comment]
> Perhaps each node in the tree could have a checkbox in the inspector to mark it as "pickable" by the eventsheet? Additionally, objects created by the eventsheet could have this option enabled by default. Just brainstorming ways to handle potential challenges.

I have plans to do something similar, but I haven't done it yet. I made a temporary solution that simply takes objects from the scene to see how it will look in the event and event adding window. The idea is that I created 2 objects "VNode2D" and "VNode3D" that store "event_sheet_data" from which the script will be generated and executed specifically on this object. Perhaps over time I will change or improve the approach to this, but for now this is it.

**Comment by prominentdetail** [comment]
This is looking very nice. It has been a while since I used construct, but I can definitely see the similarity. It has a familiar, comfortable feeling to me, and hope to see more progress!

**Comment by WladekProd** [comment]
![image](https://github.com/user-attachments/assets/ae5b6514-3b1b-4bad-9acc-e99421411985)

https://github.com/user-attachments/assets/39250bd2-bbb6-4663-ba92-0e7bbafc8bde

**Comment by blurymind** [comment]
This is very cool, but wouldnt you prefer something that requires less steps? I think Gdevelop is a better example of this being streamlined where you can add items without even opening the wizard and its a single step.
When you use the dialog way - it has a search field that takes you directly to what you need and its a single screen to set up, not a wizard like construct's
